[2.0.x] Cleanup after SD Print Again, whitespace, tabs

thinkyhead opened the pull request:
Followup to #8106

- Move LCD-oriented code from `cardreader.cpp` to `utlralcd.cpp`
- ~~Add a test commit to with code that's _supposed_ to work~~

thinkyhead committed:
Move SD_REPRINT_LAST_SELECTED_FILE to ultralcd.cpp

Roxy-3D commented:
```cpp
      void lcd_reselect_last_file() {
        // Some of this is a hack to force the screen update to work.
        // TODO: Fix the real issue that causes this!
```
I have some more information on this.   (I'm still not at the finish line yet.)   When dropping into the SD File menu, it is now easier to get the 'select error' to happen.   If the user turns the dial fast, it is possible for the previously selected file to not get entirely 'deselected' on the LCD screen.

I'm pretty sure this 'deselect issue' is also related to the incremental screen updates where only 1/2 of the screen is updated per pass.    I don't know that for a fact yet.   But I think that is true.   And this 'deselect issue' appears to be related to the screen not getting repainted also.

thinkyhead commented:
I'm also curious about the 5s timer you've added to the SD file menu. It looks like every 5 seconds it will move the current selection somewhere else, regardless of what the user may be doing. I would think that the `lcd_goto_screen(lcd_sdcard_menu, saved_encoderPosition);` call would take care of moving the selection to the same line, as it does with "BACK" items.

I also notice that with the `SD_REPRINT_LAST_SELECTED_FILE` option enabled, the SD menu will never time out anymore. Maybe it should timeout in the usual case, but not timeout when re-selecting the file for print again. The set of `defer_return_to_status` in `lcd_reselect_last_file` should take care of that, making it unnecessary in `lcd_sdcard_menu`.

thinkyhead commented on Marlin/src/lcd/ultralcd.cpp:
```diff
@@ -3756,8 +3771,8 @@ void kill_screen(const char* lcd_msg) {
-        saved_encoderPosition = encoderPosition;
-        defer_return_to_status = true;
+        //saved_encoderPosition = encoderPosition; // saved when print starts
+        //defer_return_to_status = true;           // already done in lcd_reselect_last_file
```
Looks like these may be obsolete but forgot to remove?

thinkyhead commented on Marlin/src/lcd/ultralcd.cpp:
```diff
@@ -3742,22 +3742,37 @@ void kill_screen(const char* lcd_msg) {
-        saved_encoderPosition = encoderPosition;
-        defer_return_to_status = true;
+        //saved_encoderPosition = encoderPosition; // saved when print starts
+        //defer_return_to_status = true;           // already done in lcd_reselect_last_file
```
Looks like these may be obsolete but forgot to remove?

thinkyhead commented on Marlin/src/lcd/ultralcd.cpp:
```diff
@@ -3742,22 +3742,37 @@ void kill_screen(const char* lcd_msg) {
 
       #if ENABLED(SD_REPRINT_LAST_SELECTED_FILE)
+        static millis_t assume_print_finished = 0;
         if (ELAPSED(millis(), assume_print_finished)) { // if the printer has been busy printing, lcd_sdcard_menu() should not
```
Note that this will fire always right away if you leave the SD menu for 5s or more and then return to it again.

thinkyhead commented on Marlin/src/sd/cardreader.cpp:
```diff
@@ -922,15 +913,7 @@ void CardReader::printingHasFinished() {
-      lcd_goto_screen(lcd_sdcard_menu, saved_encoderPosition);
-      defer_return_to_status = true;
-      lcd_update();
+      lcd_reselect_last_file();
```
Better to just make a function so `cardreader.h` doesn't need to know so much about the LCD vars.

thinkyhead commented:
> related to the screen not getting repainted also.

The screen handler needs to be called once per stripe, but the screen handler should only do processing (move the cursor, etc.) on the first stripe, otherwise the next stripe may be displaying different info (e.g., a different menu item selected, etc.). We had a patch to fix this before, so maybe it got lost in a later update.

thinkyhead commented on Marlin/src/lcd/ultralcd.cpp:
```diff
@@ -3742,22 +3742,37 @@ void kill_screen(const char* lcd_msg) {
           lcdDrawUpdate = LCDVIEW_REDRAW_NOW;           // have been active for 5 seconds.  In this case, restore the previous
           encoderPosition = saved_encoderPosition;      // encoderPosition to the last selected item.
           assume_print_finished = millis() + 5000;
         }
```
Note that this will fire always right away if you leave the SD menu for 5s or more and then return to it again, and will fire every 5s while inside the SD menus.

If it only needs to fire the first time when returning to the SD menu after a print, instead of a timer, a simple flag set on Start SD Print should be sufficient.

thinkyhead replied:
I'll have to test and see what happens if you remove the whole block from the function. Seems like everything should still work… Also, it seems like it would be best in `lcd_reselect_last_file` to call `lcd_goto_screen` at the very top, and only afterward force at least one screen update.

thinkyhead commented on Marlin/src/lcd/ultralcd.cpp:
```diff
@@ -3742,22 +3742,37 @@ void kill_screen(const char* lcd_msg) {
+        lcd_goto_screen(lcd_sdcard_menu, saved_encoderPosition);
+        defer_return_to_status = true;
+        lcd_update();
+      }
```
It seems this could call `lcd_goto_screen` at the very top, and only afterward force at least one screen update.

thinkyhead commented on Marlin/src/lcd/ultralcd.cpp:
```diff
@@ -3742,22 +3742,37 @@ void kill_screen(const char* lcd_msg) {
+        safe_delay(50);
+        _lcd_synchronize();
+        lcdDrawUpdate = LCDVIEW_CALL_REDRAW_NEXT;
+        drawing_screen = screen_changed = true;
```
`screen_changed` is always set by `lcd_goto_screen`. It acts as a cue for menu items following the one that activated the screen change, telling them to not activate their actions too. Clearing `lcd_clicked` would likely have the same effect, but apparently we wanted to preserve `lcd_clicked` for items following the action item.

thinkyhead commented:
I've added an experimental extra commit going back to basics on what is _supposed_ to work, with one hack, which is to use `LCDVIEW_CLEAR_CALL_REDRAW` after `lcd_goto_screen` to get the screen to do a full refresh.

I have a graphical display on the CR-10 here at MakerArm offices, plus plenty of character displays, so I'll give this all a test in the morning!

AnHardt commented:
Actual screen changing could be done in a function `change_screen_now()`, protected by the stripe clipping functionality. (either first or last stripe)
`want_to_change_screed()` could be used everywhere.

Roxy-3D commented:
> I'm also curious about the 5s timer you've added to the SD file menu. It looks like every 5 seconds it will move the current selection somewhere else, regardless of what the user may be doing. I would think that the lcd_goto_screen(lcd_sdcard_menu, saved_encoderPosition); call would take care of moving the selection to the same line, as it does with "BACK" items.

```cpp
      #if ENABLED(SD_REPRINT_LAST_SELECTED_FILE)
        if (ELAPSED(millis(), assume_print_finished)) { // if the printer has been busy printing, lcd_sdcard_menu() should not 
          lcdDrawUpdate = LCDVIEW_REDRAW_NOW;           // have been active for 5 seconds.  In this case, restore the previous
          encoderPosition = saved_encoderPosition;      // encoderPosition to the last selected item.
          assume_print_finished = millis() + 5000;
        }
        saved_encoderPosition = encoderPosition;
        defer_return_to_status = true;
      #endif      
```
The 'right' way to look at it is if this menu hasn't been called for 5 seconds, the next time through it will grab the value from saved_encoderPosition and reset the timer.    But then, either way, it will resave the current encoder position in saved_encoderPosition.

> I also notice that with the SD_REPRINT_LAST_SELECTED_FILE option enabled, the SD menu will never time out anymore. Maybe it should timeout in the usual case, but not timeout when re-selecting the file for print again. The set of defer_return_to_status in lcd_reselect_last_file should take care of that, making it unnecessary in lcd_sdcard_menu.

Maybe...   But most likely somebody using this feature isn't going to want time outs anyway.  I don't know.    But the set of defer_return_to_status does **_NOT_** stop the switch of screens.   That is why the extra logic here:
```cpp
      // Return to Status Screen after a timeout
      if (currentScreen == lcd_status_screen || defer_return_to_status)
        #if ENABLED(SD_REPRINT_LAST_SELECTED_FILE)
          if (currentScreen != lcd_sdcard_menu)                // lcd_sdcard_menu() does not time out if ENABLED(SD_REPRINT_LAST_SELECTED_FILE)
            return_to_status_ms = ms + LCD_TIMEOUT_TO_STATUS;  // When the printer finishes a file, it will wait with the file selected for 
        #else                                                  // a re-print.
        return_to_status_ms = ms + LCD_TIMEOUT_TO_STATUS;
        #endif
      else if (ELAPSED(ms, return_to_status_ms))
        lcd_return_to_status();
```
> The screen handler needs to be called once per stripe, but the screen handler should only do processing (move the cursor, etc.) on the first stripe, otherwise the next stripe may be displaying different info (e.g., a different menu item selected, etc.). We had a patch to fix this before, so maybe it got lost in a later update.

I did not see anything that keeps the position from moving on the first paint of the first stripe.  Maybe it is there.   But I did not see it.   We might want to see if we can find this previous 'fix'.

Part of the complexity here is because once again, the LCD code is being controlled from outside of its world.    The LCD code did not envision that it should be a slave to the main firmware body.

> Actual screen changing could be done in a function change_screen_now(), protected by the stripe clipping functionality. (either first or last stripe)  want_to_change_screen() could be used everywhere.

Yes.   Some idea like this could help untangle the behavior.

Roxy-3D commented:
> I've added an experimental extra commit going back to basics on what is supposed to work, with one hack, which is to use LCDVIEW_CLEAR_CALL_REDRAW after lcd_goto_screen to get the screen to do a full refresh.

I did try using LCDVIEW_CLEAR_CALL_REDRAW initially.   It didn't work.   But I'll try it with these other changes and see what it does.   It will be a bit before I can try it.  My printer is busy.   

And we need to revert these changes:   https://github.com/MarlinFirmware/Marlin/issues/7076#issuecomment-339997734       This commit broke most people's Baby-Stepping:  https://github.com/MarlinFirmware/Marlin/commit/22811540258c9d62d749446cccf262eec80523d9

thinkyhead commented:
> The 'right' way to look at it is if this menu hasn't been called for 5 seconds, the next time through it will grab the value from saved_encoderPosition and reset the timer. But then, either way, it will resave the current encoder position in saved_encoderPosition.

Yes, that describes what the code does. But, is it really needed?

thinkyhead commented:
> The 'right' way to look at it is if this menu hasn't been called for 5 seconds, the next time through it will grab the value from saved_encoderPosition and reset the timer. But then, either way, it will resave the current encoder position in saved_encoderPosition.

Yes, that describes what the code does (leaving out the part where it keeps firing every 5s). But, is it really needed?

thinkyhead commented:
> the set of defer_return_to_status does NOT stop the switch of screens

In that case it's getting cleared elsewhere, because when set it does just that. I will investigate through careful testing.

thinkyhead commented:
> The LCD code did not envision that it should be a slave to the main firmware body.

And, indeed, until the addition of recent features (M600, UBL) we were careful to only control the LCD code from within the LCD code. See the original and current implementation of `MESH_BED_LEVELING`+`LCD_BED_LEVELING`, for example.

thinkyhead commented:
> The LCD code did not envision that it should be a slave to the main firmware body.

Well, facilities can always be added to the LCD code to make it more compliant for this kind of usage, and in due time we will get around to that. At present there are a couple different methods being employed. So, those will be used as guidance.

thinkyhead commented:
> I did not see anything that keeps the position from moving on the first paint of the first stripe.

The menu line will only change on the first call of a menu handler, because `encoderPosition` gets cleared by that first call. The click flag is also cleared on the first invocation. So, subsequent calls should not move the cursor or detect a click.

thinkyhead commented:
> I did not see anything that keeps the position from moving on the first paint of the first stripe.

The menu line will only change on the first call of a menu handler, because `encoderPosition` gets cleared by that first call. The click flag is also cleared on the first invocation. So, subsequent calls should not move the cursor or detect a click. The important thing is for events to only be dispatched when calling the screen handler for the first stripe ("`page.page == 0`").

thinkyhead commented:
Ideally, it would be nice to be able to separate the processing part from the drawing part, but the menu code is too clever for its own good. Creating an entirely new LCD screen/menu API will be a challenge, but going "from the ground-up" seems like a very good approach. Actually developing a design, with requirements… It's something to start in a branch of its own …… along with that 32-bit Marlin from scratch ……

Roxy-3D commented:
>> The 'right' way to look at it is if this menu hasn't been called for 5 seconds, the next time through it will grab the value from saved_encoderPosition and reset the timer. But then, either way, it will resave the current encoder position in saved_encoderPosition.

> Yes, that describes what the code does (leaving out the part where it keeps firing every 5s). But, is it really needed?

I'll review and have more comments in the morning....   (its Friday and I've already been playing darts.)   But here is the deal...  Part of the problem is it is not firing every 5 seconds.  Even when the person is sitting in that menu...   The menu is **_NOT_** being called.   It is only with an encoder wheel turn that anything happens....   That is a big part of the problem.   There are no  CPU cycles being allocated there.

Roxy-3D commented:
>> the set of defer_return_to_status does NOT stop the switch of screens

> In that case it's getting cleared elsewhere, because when set it does just that. I will investigate through careful testing.

I think defer_return_to_status does not operate like you think it does.   It literally 'defers'.  It does not stop a return to status.   Maybe we should change this line to:       return_to_status_ms = ms + 9999999; ???
```cpp
    #if ENABLED(ULTIPANEL)
      // Return to Status Screen after a timeout
      if (currentScreen == lcd_status_screen || defer_return_to_status)
        return_to_status_ms = ms + LCD_TIMEOUT_TO_STATUS;
      else if (ELAPSED(ms, return_to_status_ms))
        lcd_return_to_status();
    #endif // ULTIPANEL
```

> Ideally, it would be nice to be able to separate the processing part from the drawing part, but the menu code is too clever for its own good. Creating an entirely new LCD screen/menu API will be a challenge, but going "from the ground-up" seems like a very good approach. Actually developing a design, with requirements… It's something to start in a branch of its own …… along with that 32-bit Marlin from scratch ……

Yes.   And I'm certainly not saying we need to do it immediately.   But it is important to recognize many of the problems we are seeing are because code outside of the LCD Status and menu system are trying to use the LCD Panel for their own reasons.     What ever ideas we have for a 'clean up' should factor that into the discussion.

thinkyhead commented:
> I think `defer_return_to_status` does not operate like you think it does.

Oh ye of little faith. You've misunderstood the code you cited. It shows `defer_return_to_status` working exactly as I described and as I specified the flag to work when I first added it. When set, `return_to_status_ms` is perpetually updated (also does this when the status screen is active). Also, when `defer_return_to_status` is set, it **_cannot_** get to the code that does the return to status:
```cpp
      else if (ELAPSED(ms, return_to_status_ms))
        lcd_return_to_status();
```

thinkyhead commented:
Here's what your added code does to break it:
```cpp
      if (currentScreen == lcd_status_screen || defer_return_to_status)
          if (currentScreen != lcd_sdcard_menu)                // lcd_sdcard_menu() does not time out if ENABLED(SD_REPRINT_LAST_SELECTED_FILE)
            return_to_status_ms = ms + LCD_TIMEOUT_TO_STATUS;  // When the printer finishes a file, it will wait with the file selected for
      else if (ELAPSED(ms, return_to_status_ms))
        lcd_return_to_status();
```
The solution is to add braces so the new `if` statement doesn't attach to the following `else`.

thinkyhead commented:
Here's what your added code does to break it:
```cpp
// Return to Status Screen after a timeout
if (currentScreen == lcd_status_screen || defer_return_to_status)
  if (currentScreen != lcd_sdcard_menu)
    return_to_status_ms = ms + LCD_TIMEOUT_TO_STATUS;
  else if (ELAPSED(ms, return_to_status_ms))
    lcd_return_to_status();
```
The solution is to add braces so the new `if` statement doesn't attach to the following `else`.
```cpp
// Return to Status Screen after a timeout
if (currentScreen == lcd_status_screen || defer_return_to_status) {
  #if ENABLED(SD_REPRINT_LAST_SELECTED_FILE)
    if (currentScreen != lcd_sdcard_menu)                // lcd_sdcard_menu() does not time out if ENABLED(SD_REPRINT_LAST_SELECTED_FILE)
      return_to_status_ms = ms + LCD_TIMEOUT_TO_STATUS;  // When the printer finishes a file, it will wait with the file selected for
  #else                                                  // a re-print.
    return_to_status_ms = ms + LCD_TIMEOUT_TO_STATUS;
  #endif
}
else if (ELAPSED(ms, return_to_status_ms))
  lcd_return_to_status();
```
…so… if that was your intention…
```cpp
// Return to Status Screen after a timeout
if (currentScreen == lcd_status_screen || defer_return_to_status) {
  if (currentScreen != lcd_sdcard_menu)
    return_to_status_ms = ms + LCD_TIMEOUT_TO_STATUS;
}
else if (ELAPSED(ms, return_to_status_ms))
  lcd_return_to_status();
```
But this isn't logical either.

So, just use `defer_return_to_status`. It works!

thinkyhead committed:
Remove return_to_status_ms hack, just use defer...

thinkyhead committed:
Remove 5s timer hack. Encoder pos is saved on print start!

thinkyhead commented:
Tested and working with proposed cleanup (character display). As the extra redraw hack is still left in for graphical display, it should also work on DOGLCD.

thinkyhead commented:
- Removed 5s timer code from `lcd_sdcard_menu`
- Removed code from `lcd_update` and trust `defer_return_to_status`
- Made redraw hack in `lcd_reselect_last_file` DOGLCD-only

Tested and working with proposed cleanup (character display). As the extra screen refresh hack is still left in for graphical display, it should still work on DOGLCD.

Again, the screen refresh hack may not be needed if there's a general fix for broken screen changes that can be applied.

This new feature's entire upshot is that when a print ends, there's a screen change. If that screen change is failing, then other screen changes should frequently be failing too. Continue to strip out the DOGLCD redraw hack from `lcd_reselect_last_file` until you find the minimum needed hack, and that will give us a clue where to look to fix screen change code.

thinkyhead commented:
I'll post a brief post-mortem on this feature to show you where bad practices were followed, and how to avoid them in the future.

thinkyhead commented:
Added code to prevent completion of `auto#.g` from triggering Print Again.
